api/proto/teleport/legacy/client/proto/authservice.proto

```diff
@@ -3290,6 +3310,22 @@ service AuthService {
   // DeleteRole deletes an existing role in a backend described by the given request.
   rpc DeleteRole(DeleteRoleRequest) returns (google.protobuf.Empty);
 
+  // AtomicWriteRoles performs atomic role create/update/delete operations using Compare-And-Swap.
+  //
+  // All operations succeed or fail together as a single transaction. This prevents partial state changes
+  // in critical role management scenarios.
+  //
+  // Roles in roles_to_create are created with NotExists() condition, roles in roles_to_update are updated
+  // with Revision() condition for optimistic concurrency control, and roles in role_names_delete are
+  // deleted using Exists() condition.
+  //
+  // Use cases: For transactional UI operations where a single user action triggers multiple role changes,
+  // atomicity ensures the backend state matches user intent, guaranteeing that either the entire operation
+  // succeeds or nothing changes and the user can safely retry. Without atomicity, partial failures can
+  // leave the system in an inconsistent state requiring manual intervention to identify and fix incomplete
+  // operations.
+  rpc AtomicWriteRoles(AtomicWriteRolesRequest) returns (AtomicWriteRolesResponse);
```

Comment:
I don't know that we should provide a generic API to atomically modify an unlimited number of roles. I'm also not sure that we should be exposing backend concepts like AtomicWrite in RPC names. 

Can we change the API to be tailored to this use case instead of a generic one size fits all transaction? Maybe something like ModifyRolesForAccessList which can then internally use AtomicWrite today or any other transactional mechanism the backend might provide tomorrow?

> In the RFD 0236 main use case we have 3 roles to created. The AtomicWriteRoles can be limited to 10 roles for now extended if needed.

Does the use case require updating and deleting roles? What is the maximum number of roles we'd expect to update and delete?